### ISSUE TITLE
[Bug] FA2 attention silently outputs zeros when every logit in a row is below -5e4 (math::inf sentinel is a finite value)

### ISSUE THREAD

**Issue by tangcy98** [flashinfer-ai/flashinfer#4267]
### Before submitting

- [x] I have searched existing issues and did not find a solution or an existing report.

### Environment

```text
### FlashInfer environment report
<!-- generated by `python -m flashinfer.collect_env`; paste into your issue -->

==== FlashInfer ====
flashinfer                   : 0.6.12
flashinfer file              : /usr/local/lib/python3.12/dist-packages/flashinfer/__init__.py
flashinfer-cubin             : 0.6.12
flashinfer-jit-cache         : 0.6.12+cu130  ⚠ MISMATCH vs flashinfer-python==0.6.12
JIT cache                    : /root/.cache/flashinfer (0 compiled .so)
Cubin store (local)          : /usr/local/lib/python3.12/dist-packages/flashinfer_cubin/cubins (15558 cubins, 1754 MiB)
Target CUDA archs (resolved) : {(12, '0f')}

==== Python / Platform ====
Python            : 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
Python executable : /usr/bin/python
Platform          : Linux-5.15.0-181-generic-x86_64-with-glibc2.39
libc              : glibc 2.39
OS                : Ubuntu 24.04.3 LTS
Container         : no / not detected

==== GPU / Driver ====
Driver version       : 580.126.20
CUDA_VISIBLE_DEVICES : <unset>
GPU 0 (CUDA order)   : NVIDIA RTX PRO 5000 72GB Blackwell | SM120 | 110 SMs | 71.1 GiB
GPU 1 (CUDA order)   : NVIDIA RTX PRO 5000 72GB Blackwell | SM120 | 110 SMs | 71.1 GiB

==== CUDA Toolkit ====
nvcc on PATH                            : not found
CUDA_HOME (env)                         : <unset>
CUDA toolkit resolved by flashinfer JIT : /usr/local/cuda (version 13.0)

==== PyTorch ====
torch                          : 2.11.0+cu130
torch.version.cuda             : 13.0
torch CXX11 ABI                : True
torch compiled archs           : sm_75 sm_80 sm_86 sm_90 sm_100 sm_120
torch.backends.cudnn.version() : 91900
torch file                     : /usr/local/lib/python3.12/dist-packages/torch/__init__.py

==== cuDNN Frontend ====
cudnn-frontend                       : 1.26.0
cudnn backend (loaded, via frontend) : 9.19.0
cudnn-frontend file                  : /usr/local/lib/python3.12/dist-packages/cudnn/__init__.py

==== GPU Libraries: loaded vs on disk ====
libcudnn  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/lib/python3.12/dist-packages/nvidia/cudnn/lib/  [v9]
    on disk  /usr/lib/x86_64-linux-gnu/  [v9.13.0]
libcublas  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/lib/python3.12/dist-packages/nvidia/cu13/lib/  [v13]
    on disk  /usr/local/cuda-13.0/targets/x86_64-linux/lib/  [v13.0.2.14]
libcudart  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/lib/python3.12/dist-packages/nvidia/cu13/lib/  [v13]
    on disk  /usr/local/cuda-13.0/targets/x86_64-linux/lib/  [v13.0.88]
libnvrtc  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/lib/python3.12/dist-packages/nvidia/cu13/lib/  [v13]
    on disk  /usr/local/cuda-13.0/targets/x86_64-linux/lib/  [v13.0.88]
libnccl  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/lib/python3.12/dist-packages/nvidia/nccl/lib/  [v2]
    on disk  /usr/lib/x86_64-linux-gnu/  [v2.28.3]
libcuda (driver):
    LOADED   /usr/lib/x86_64-linux-gnu/  [v580.126.20]

==== Relevant Packages ====
apache-tvm-ffi               : 0.1.11
cuda-bindings                : 13.2.0
cuda-core                    : 1.0.1
cuda-pathfinder              : 1.5.6
cuda-python                  : 13.2.0
cuda-tile                    : 1.3.0
cuda-toolkit                 : 13.0.2
flashinfer-cubin             : 0.6.12
flashinfer-jit-cache         : 0.6.12+cu130
flashinfer-python            : 0.6.12
ninja                        : 1.13.0
numpy                        : 2.3.5
nvidia-cublas                : 13.1.0.3
nvidia-cuda-crt              : 13.3.73
nvidia-cuda-cupti            : 13.0.85
nvidia-cuda-nvcc             : 13.2.78
nvidia-cuda-nvrtc            : 13.0.88
nvidia-cuda-runtime          : 13.0.96
nvidia-cuda-tileiras         : 13.2.78
nvidia-cudnn-cu13            : 9.19.0.56
nvidia-cudnn-frontend        : 1.26.0
nvidia-cufft                 : 12.0.0.61
nvidia-cufile                : 1.15.1.6
nvidia-curand                : 10.4.0.35
nvidia-cusolver              : 12.0.4.66
nvidia-cusparse              : 12.6.3.3
nvidia-cusparselt-cu13       : 0.8.0
nvidia-cutlass-dsl           : 4.5.2
nvidia-cutlass-dsl-libs-base : 4.5.2
nvidia-cutlass-dsl-libs-cu13 : 4.5.2
nvidia-mathdx                : 25.6.0
nvidia-ml-py                 : 13.610.43
nvidia-modelopt              : 0.45.0
nvidia-nccl-cu13             : 2.28.9
nvidia-nvjitlink             : 13.0.88
nvidia-nvshmem-cu13          : 3.4.5
nvidia-nvtx                  : 13.0.85
nvidia-nvvm                  : 13.2.78
sglang                       : 0.5.15+minicpm
sglang-kernel                : 0.4.4
sglang-router                : 0.3.2
torch                        : 2.11.0+cu130
torch-c-dlpack-ext           : 0.1.5
torch-memory-saver           : 0.0.9.post1
torchao                      : 0.17.0+cu130
torchaudio                   : 2.11.0+cu130
torchcodec                   : 0.11.1+cu130
torchvision                  : 0.26.0+cu130
transformers                 : 5.12.1
triton                       : 3.6.0

==== Environment Variables ====
PATH : /root/.local/bin/:/root/.cursor-server/bin/linux-x64/55434bd8062ece6fee083b82beed2aee42d253f0/bin/remote-cli:/root/.cursor-server/bin/linux-x64/55434bd8062ece6fee083b82beed2aee42d253f0/bin/remote-cli:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

==== GPU Topology ====
nvidia-smi topo -m:
GPU0    GPU1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NODE    128-255 1               N/A
GPU1    NODE     X      128-255 1               N/A
```

### Bug description

Summary. The fa2 attention kernels initialize the online-softmax running max m to -math::inf, where math::inf is a finite value of 5e4 (include/flashinfer/math.cuh:33 on main; :32 on the 0.6.12 tag). When every real logit in a row falls below that sentinel, m = max(m, rowmax) stays clamped at -5e4, every exp2() term underflows to zero, and the kernel silently returns a zero vector for that row. No error, no NaN — just a wrong answer.

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/math.cuh#L33

How we hit it. A bf16 GQA model we serve produces degenerate repetitive text on longer generations with the flashinfer backend, while --attention-backend triton on an identical build is correct. The model has massive activations in its final attention layer, with |q|max ≈ 448 and |k|max ≈ 1880 driving raw pre-sm_scale QK logits down to -2.45e5. One head has a row whose logits are all large negative — that head's output is exactly zero on every decode step, corrupting the final hidden state.

Massive activations of this kind are not exotic; any model without QK-norm can develop them, so we expect this to affect other deployments silently.

Mechanism (real numbers, head_dim = 128, sm_scale = 0.088388, so sm_scale_log2 = 0.1275):

m_prev  = -5e4                                     // prefill.cuh:901, initial sentinel
rowmax  = -2.45e5                                  // real logits, all below the sentinel
m       = max(-5e4, -2.45e5) = -5e4                 // clamped, never moves
o_scale = exp2((m_prev - m) * sm_scale_log2) = 1
s_frag  = exp2((-2.45e5 + 5e4) * 0.1275) = 0        // every term underflows
d       = 1.0                                       // stays at its init value (prefill.cuh:901)
o_frag  = 0

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/prefill.cuh#L896-L905

o_frag is already zero and d is still 1.0 before the epilogue runs, so the output is a zero vector regardless of which branch OutputTransform takes.

Note on the sentinel comparison. variant_helper.cuh:86 does float d_rcp = (m != -math::inf) ? math::ptx_rcp(d) : 0.f; — it uses "m still equals the sentinel" as the test for "this row was fully masked". That test also misfires here, but as shown above it is not what produces the zeros: the underflow in update_mdo_states already did. Replacing the equality test with an explicit "row fully masked" flag would therefore not fix this bug — the sentinel value itself has to move.

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/variant_helper.cuh#L85-L89

A second, nastier variant of the same root cause. KTraits::MaskFillValue is also -math::inf = -5e4 (prefill.cuh:339/345, and AttentionUpdater::fill_value at hopper/attention_updater.cuh:168). If a row contains both masked/padding positions and real logits below -5e4, the masked positions become the row maximum and receive softmax weight ≈ 1. The output then degenerates into an average of the V rows at masked positions — a non-zero garbage value rather than a clean zero, which is considerably harder to notice. We observed exact zeros only because the corresponding V entries happened to be zero.

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/prefill.cuh#L339-L347

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/hopper/attention_updater.cuh#L167-L168

Threshold is in the raw (pre-sm_scale) domain for fa2. In prefill.cuh the scale is applied inside exp2 (prefill.cuh:1577) and m is only multiplied by sm_scale_log2 later in finalize_m (prefill.cuh:1754), so m holds unscaled logits throughout and the breakpoint is a sm_scale-independent raw -5e4. This matches our measured breakpoint exactly: scaled -4450 at sm_scale = 0.088388 → raw -5.03e4.

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/prefill.cuh#L1576-L1585

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/prefill.cuh#L1753-L1757

Inconsistency between the two paths. decode.cuh:94-99 applies s[j] *= variant.sm_scale_log2 before comparing against -math::inf, so on that path the sentinel lives in the scaled domain and the effective raw threshold becomes -5e4 / (sm_scale * log2e) instead. The same named constant serves as a sentinel in two different domains. This explains an otherwise confusing observation in our sweeps: at sm_scale = 1.0 the breakpoint moved to roughly -3.47e4 rather than staying at -5e4.

https://github.com/flashinfer-ai/flashinfer/blob/4b964ec4e147cf06a39e08b08c43188859df2652/include/flashinfer/attention/decode.cuh#L93-L95

Expected behavior. A row of finite, well-ordered logits should produce a correct softmax-weighted output regardless of its absolute magnitude. -2.45e5 is far inside fp32 range and the row has an unambiguous argmax (top1/top2 gap = 4545, vs ~0.003 fp32 resolution at that magnitude).

Affected paths. fa2 prefill and fa2 tensor-core decode (which reuses the prefill kernel). fa3 appears immune: hopper/attention_updater.cuh:185 initializes row_max from the real scores via reduce_max</*init=*/true> and finalize has no sentinel comparison. We could not verify this by running, since fa3 is unavailable on sm120.

### Minimal reproduction

```python
# Requires only flashinfer + torch. No model weights.
# Sweeps the row-max of an all-negative logit row across the -5e4 sentinel.
import torch
import flashinfer

dev = "cuda"
D, HQ, HKV, KV = 128, 16, 1, 64   # head_dim / qo heads / kv heads / kv len

def run(level: float):
    """Build q/k so every raw logit (q·k, before sm_scale) == `level`,
    except key 0 which is an unambiguous argmax at `level + 5000`."""
    q = torch.full((1, HQ, D), 1.0, dtype=torch.bfloat16, device=dev)
    k = torch.full((KV, HKV, D), level / D, dtype=torch.bfloat16, device=dev)
    k[0].fill_((level + 5000.0) / D)
    v = torch.randn(KV, HKV, D, dtype=torch.bfloat16, device=dev)

    o = flashinfer.single_prefill_with_kv_cache(q, k, v, causal=False, backend="fa2")

    # fp32 reference; sdpa's default scale 1/sqrt(D) matches flashinfer's default sm_scale
    ref = torch.nn.functional.scaled_dot_product_attention(
        q.transpose(0, 1).unsqueeze(0).float(),
        k.transpose(0, 1).unsqueeze(0).float().expand(1, HQ, KV, D),
        v.transpose(0, 1).unsqueeze(0).float().expand(1, HQ, KV, D),
    ).squeeze(0).transpose(0, 1)

    rel = (o.float() - ref).norm() / ref.norm()
    print(f"raw row-max = {level + 5000:>11.0f}   rel_l2 = {rel:.5f}   |out| = {o.float().norm():.4f}")

for level in [-1e4, -3e4, -5e4, -6e4, -1e5, -2.45e5]:
    run(level)

Expected: rows whose maximum stays above raw -5e4 are accurate (rel_l2 ~ 1e-3); once the row maximum drops below -5e4, rel_l2 jumps to 1.0 and |out| becomes exactly 0. Running with backend="fa2" on an sm90 machine should reproduce identically; backend="fa3" should not.
```

### Error messages / logs

```text
raw row-max =       -5000   rel_l2 = 0.00000   |out| = 42.4961
raw row-max =      -25000   rel_l2 = 0.00000   |out| = 46.9097
raw row-max =      -45000   rel_l2 = 0.00000   |out| = 50.0367
raw row-max =      -55000   rel_l2 = 1.00000   |out| = 0.0000
raw row-max =      -95000   rel_l2 = 1.00000   |out| = 0.0000
raw row-max =     -240000   rel_l2 = 1.00000   |out| = 0.0000

No exception is raised and no warning is emitted — this is a silent numerical failure. That is the main hazard: short prompts are correct under both backends, so the bug survives casual smoke tests. It only becomes visible in longer generations, where the per-step error accumulates.
```

**Comment by shoutoutuoadi325** [comment]
I've investigated this issue and opened a PR to fix it.

**Summary**: The bug is caused by  (a finite sentinel) in `include/flashinfer/math.cuh`. When valid logits fall below -50000, the running max clamps at this sentinel, causing exp() underflow and zero outputs.

**Fix**: Changed to true IEEE infinity: `std::numeric_limits<float>::infinity()`

**PR**: https://github.com/flashinfer-ai/flashinfer/pull/4401

The fix is minimal (1 line changed), principled (matches the pattern used in sampling/topk modules), and includes comprehensive regression tests covering logits from -1e4 to -2.45e5.

**Environment**: RTX 4090 / SM89, CUDA 12.1, PyTorch 2.5.1+cu121

**Comment by anguyen8** [comment]
B200 follow-up matrix, using deterministic standalone scripts and the same PyTorch/CUDA environment for every build:

| Regression surface | `v0.6.16.post3` | `main@2fb785c` | `#4401@2553332` |
|---|---|---|---|
| #4450 classic paged decode, two valid keys | output `0`, sentinel LSE | output `0`, sentinel LSE | **passes**: output `1`, reference LSE |
| #4451 ragged FA2, one valid key | output `0`, sentinel LSE | output `0`, sentinel LSE | **fails with NaN**: 4,096/4,096 output values and 32/32 LSE values |
| #4452 paged FA2, 1,056 fully masked head-rows | output `0`, finite sentinel LSE | output `0`, finite sentinel LSE | **fails with NaN**: all 135,168 empty-row output values and all 1,056 empty-row LSE values |

This confirms the finite `5e4` sentinel mechanism is still present on current main and that the true-infinity change in #4401 is sufficient for the classic paged-decode fixture. It is not sufficient for the ragged one-key or genuinely empty paged-prefill fixtures: those change from zero/sentinel results to NaN on the tested PR head.

Each linked issue now contains its full deterministic script, actual stdout/traceback, public source references, and untrimmed `python -m flashinfer.collect_env` output for all three builds.
